### PR TITLE
add detection output operator for supporting retinanet

### PR DESCRIPTION
**Pull request by FlyingQianMM**
test=develop

This operation is to get the detection results by performing following steps: (1)Divide the input bounding box predictions into box predictions per FPN level, according to that the box number of the i-th FPN level is 4**(MAX_LEVEL-i) times of that of the MAX_LEVEL-th FPN level. (2) Decode top-scoring bounding box predictions per FPN level according to the anchor boxes. (3)Merge top predictions from all levels and applying multi-class non maximum suppression (NMS) on them to get the final detections.
image

### CODE REVIEW

**Review by heavengate — DISMISSED** [review]

**Review by heavengate — DISMISSED** [review]

File: paddle/fluid/operators/detection/retinanet_detection_output_op.cc [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+limitations under the License. */
+
+#include <glog/logging.h>
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/operators/detection/poly_util.h"
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/retinanet_detection_output_op.cc]: 这个有用到poly里的函数吗？

File: paddle/fluid/operators/detection/retinanet_detection_output_op.cc [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+      auto* im_info_data = im_info.data<T>();
+      auto im_height = im_info_data[0];
+      auto im_width = im_info_data[1];
+      auto im_scale = im_info_data[2];
+      DeltaScoreToPrediction(bboxes_data, anchors_data, im_height, im_width,
+                             im_scale, class_num, sorted_indices, &preds);
+    }
+
+    MultiClassNMS(preds, class_num, keep_top_k, nms_threshold, nms_eta,
+                  nmsed_out, num_nmsed_out);
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/retinanet_detection_output_op.cc:368]: 1. 这个op和multiclass_nms_op大量重复的代码，不方便后续维护。
2. 这个解码，能用类似detection_output那样在python里用box_encode和multiclass_nms组合吗？

File: python/paddle/fluid/tests/unittests/test_retinanet_detection_output.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        num_det = keep_top_k
+    nmsed_outs = []
+    for s, c, idx in sorted_score_index:
+        xmin = prediction[c][idx][0]
+        ymin = prediction[c][idx][1]
+        xmax = prediction[c][idx][2]
+        ymax = prediction[c][idx][3]
+        nmsed_outs.append([c, s, xmin, ymin, xmax, ymax])
+
+    return nmsed_outs, num_det
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_retinanet_detection_output.py:58]: 同样，python里大量和test_multiclass_nms_op.py重复的代码

File: paddle/fluid/operators/detection/retinanet_detection_output_op.cc [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+oh = exp(phv * th) * ph + th / 2
+
+where `tx`, `ty`, `tw`, `th` denote the predicted box's center coordinates, width
+and height respectively. Similarly, `px`, `py`, `pw`, `ph` denote the
+anchor's center coordinates, width and height. `pxv`, `pyv`, `pwv`,
+`phv` denote the variance of the anchor box and `ox`, `oy`, `ow`, `oh` denote the
+decoded coordinates, width and height. 
+
+Then the top decoded prediction from all levels are merged followed by NMS.
+In the NMS step, this operator pruns away boxes that have high IOU
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/retinanet_detection_output_op.cc]: pruns?

File: paddle/fluid/operators/detection/retinanet_detection_output_op.cc [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+(intersection over union) overlap with already selected boxes by adaptive
+threshold NMS based on parameters of nms_threshold and nms_eta.
+Aftern NMS step, at most keep_top_k number of total bboxes are to be kept
```
Comment (jiangjiajun) [inline paddle/fluid/operators/detection/retinanet_detection_output_op.cc]: after

File: python/paddle/fluid/tests/unittests/test_retinanet_detection_output.py [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+import unittest
+import numpy as np
+import math
+import copy
+from op_test import OpTest
+from test_anchor_generator_op import anchor_generator_in_python
+from test_multiclass_nms_op import iou
+
+
+def nms(cls_dets, nms_threshold=0.05, eta=1.0):
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_retinanet_detection_output.py]: nms一定程度也可以复用test_multiclass_nms_op吧

File: python/paddle/fluid/tests/unittests/test_retinanet_detection_output.py [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+        self.aspect_ratios = [1.0, 2.0, 0.5]
+        self.anchor_scale = 4
+        self.anchor_strides = [8, 16, 32, 64, 128]
+
+        self.box_size = 4
+        self.class_num = 80
+        self.batch_size = 1
+        self.input_channels = 20
+
+    def init_test_input(self):
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_retinanet_detection_output.py]: 这个函数不能和上面的复用吗？

File: paddle/fluid/operators/detection/retinanet_detection_output_op.cc [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+    const T inter_ymax = std::min(box1[3], box2[3]);
+    T norm = normalized ? static_cast<T>(0.) : static_cast<T>(1.);
+    T inter_w = inter_xmax - inter_xmin + norm;
+    T inter_h = inter_ymax - inter_ymin + norm;
+    const T inter_area = inter_w * inter_h;
+    const T bbox1_area = BBoxArea<T>(box1, normalized);
+    const T bbox2_area = BBoxArea<T>(box2, normalized);
+    return inter_area / (bbox1_area + bbox2_area - inter_area);
+  }
+}
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/retinanet_detection_output_op.cc:172]: 设计上上面这些函数，传指针T*，把这些函数封装到bbox_util.h里之类的，完全可复用。重复的代码不好维护

File: paddle/fluid/operators/detection/retinanet_detection_output_op.cc [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+      RetinanetDetectionOutput(ctx, score_per_batch_list, box_per_batch_list,
+                               anchors_list, im_info_slice, &nmsed_out,
+                               &num_nmsed_out);
+      all_nmsed_out.push_back(nmsed_out);
+      batch_starts.push_back(batch_starts.back() + num_nmsed_out);
+    }
+
+    int num_kept = batch_starts.back();
+    if (num_kept == 0) {
+      T* od = outs->mutable_data<T>({1, 1}, ctx.GetPlace());
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/retinanet_detection_output_op.cc]: num_kept=0时，建议还是 outs还是为空吧：

```c++
T* od = outs->mutable_data<T>({0, 6}, ctx.GetPlace());
```

**Review by qingqing01 — DISMISSED** [review]

File: python/paddle/fluid/layers/detection.py [hunk 10]
```diff
@@ -2312,6 +2313,113 @@ def box_clip(input, im_info, name=None):
...
+       to the anchor boxes.
+    2. Merge top predictions from all levels and apply multi-class non 
+       maximum suppression (NMS) on them to get the final detections.
+
+    Args:
+        bboxes(List): A list of tensors from multiple FPN levels. Each
+            element is a 3-D Tensor with shape [N, Mi, 4] representing the
+            predicted locations of Mi bounding boxes. N is the batch size,
+            Mi is the number of bounding boxes from i-th FPN level and each 
+            bounding box has four coordinate values and the layout is
```
Comment (xsrobin) [inline python/paddle/fluid/layers/detection.py]: is->are

File: python/paddle/fluid/layers/detection.py [hunk 11]
```diff
@@ -2312,6 +2313,113 @@ def box_clip(input, im_info, name=None):
...
+        nms_top_k(int): Maximum number of detections per FPN layer to be
+            kept according to the confidences before NMS.
+        keep_top_k(int): Number of total bboxes to be kept per image after
+            NMS step. -1 means keeping all bboxes after NMS step.
+        nms_threshold(float): The threshold to be used in NMS.
+        nms_eta(float): The parameter for adaptive NMS.
+
+    Returns:
+        Variable:
+            The detection outputs is a LoDTensor with shape [No, 6].
```
Comment (xsrobin) [inline python/paddle/fluid/layers/detection.py]: outputs is->output is

**Review by xsrobin — COMMENTED** [review]
as comment

**Review by qingqing01 — APPROVED** [review]

**Review by xsrobin — APPROVED** [review]
LGTM

**Pull request merged**